### PR TITLE
chore: add ability to include custom protoc-gen-go dependency in nix flake

### LINKED ISSUE #14343: Ensure all dependencies defined in `flake.nix` and our GitHub Actions are aligned
For example, I'm having a hell of a time trying to generate some protobuf definitions which pass CI because `protoc-gen-go` is not locked in our nix flake, while it's using a fixed version in `.github/workflows/ci.yaml`.

```yaml
      - name: go install tools
        run: |
          go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.30
```

```bash
[nix-shell:~/coder]$ protoc-gen-go --version
protoc-gen-go v1.34.2
```

### PR DESCRIPTION
**Pull request by joobisb**
This PR can be a starting point to fix https://github.com/coder/coder/issues/14343

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -40,6 +40,28 @@
         # From https://nixos.wiki/wiki/Google_Cloud_SDK
         gdk = pkgs.google-cloud-sdk.withExtraComponents ([ pkgs.google-cloud-sdk.components.gke-gcloud-auth-plugin ]);
 
+        customProtocGenGo = pkgs.buildGoModule rec {
+          name = "protoc-gen-go";
+          owner = "protocolbuffers";
+          repo = "protobuf-go";
+          rev = "v1.30.0";
```
Comment (dannykopping) [inline flake.nix:47]: Hhmm I'm not seeing this version in the list?
https://pkg.go.dev/github.com/golang/protobuf/protoc-gen-go?tab=versions

File: flake.nix [hunk 2]
```diff
@@ -40,6 +40,28 @@
...
+          name = "protoc-gen-go";
+          owner = "protocolbuffers";
+          repo = "protobuf-go";
+          rev = "v1.30.0"; 
+          src = pkgs.fetchFromGitHub {
+            owner = "protocolbuffers";
+            repo = "protobuf-go";
+            rev = rev;
+            # This should be updated whenever rev changes!
+            # To update, set to "", run nix-shell, insert new hash
```
Comment (dannykopping) [inline flake.nix:53]: Where does the new hash come from?

File: flake.nix [hunk 3]
```diff
@@ -80,7 +102,7 @@
           playwright-driver.browsers
           postgresql_16
           protobuf
-          protoc-gen-go
+          customProtocGenGo
```
Comment (dannykopping) [inline flake.nix:105]: Let's follow the convention above and call it:

```suggestion
          proto_gen_go_1_30
```

**Review by dannykopping — COMMENTED** [review]
Thanks for the contribution!

File: flake.nix (hunk 2 above)
Comment (dannykopping) [reply flake.nix:53]: Can you rather integrate this into `scripts/update-flake.sh` pls?
The less we make folks think, the more likely they are to do the right thing.

Comment (joobisb) [reply flake.nix:53]: yeah that would be better, I will do that

**Comment by matifali** [comment]
I agree with the resolution to align nix and CI. But I am afraid in long run this will become a resistance to easily upgrade dev tools and packages. 

Why we can't update the version in CI in to match with nix?

Or best use nix to run our CI.

**Comment by dannykopping** [comment]
> Why we can't update the version in CI in to match with nix?

CI is the only one with a pinned version; either way we'd have to do what @joobisb has implemented.

> Or best use nix to run our CI.

That's the dream, but it'll take some time. In the interim, we can start by aligning versions IMHO.

**Comment by dannykopping** [comment]
@joobisb please avoid rebasing once a review has begun. There's no need to maintain a clean history, since we squash-merge anyway.

**Comment by joobisb** [comment]
@dannykopping I've updated the script

File: scripts/update-flake.sh [hunk 4]
```diff
@@ -13,4 +27,14 @@ echo "Calculating SRI hash..."
 HASH=$(go run tailscale.com/cmd/nardump --sri "$OUT/pkg/mod/cache/download")
 sudo rm -rf "$OUT"
 
+echo "Updating go.mod vendorHash"
 sed -i "s#\(vendorHash = \"\)[^\"]*#\1${HASH}#" ./flake.nix
+
+# Update protoc-gen-go sha256
+echo "Updating protoc-gen-go sha256..."
+PROTOC_GEN_GO_REV=$(grep -A 20 'pkgs.buildGoModule rec' flake.nix | grep -A 10 'repo = "protobuf-go"' | grep 'rev = "v' | sed 's/.*rev = "\(.*\)".*/\1/')
```
Comment (dannykopping) [inline scripts/update-flake.sh:35]: This is flimsy (and possible not portable across Linux/Mac/Windows); is there a more robust way we can get this?

**Review by dannykopping — COMMENTED** [review]
Thanks!

**Comment by joobisb** [comment]
@dannykopping I've addressed the comments, could you please have a look

**Comment by dannykopping** [comment]
@joobisb sure thing, I'm on PTO since yesterday but back tomorrow. 
Please address the CI failures as well. 
We'll get this merged soon 👍 thanks for your patience!

**Comment by matifali** [comment]
It should be called a `chore` as it's not a user-facing feature. Also, please run `make fmt` to fix the formatting issue.

File: scripts/update-flake.sh [hunk 5]
```diff
@@ -13,4 +27,14 @@ echo "Calculating SRI hash..."
...
 sudo rm -rf "$OUT"
 
+echo "Updating go.mod vendorHash"
 sed -i "s#\(vendorHash = \"\)[^\"]*#\1${HASH}#" ./flake.nix
+
+# Update protoc-gen-go sha256
+echo "Updating protoc-gen-go sha256..."
+PROTOC_GEN_GO_REV=$(nix eval --extra-experimental-features nix-command --extra-experimental-features flakes --raw .#proto_gen_go.rev)
+echo "protoc-gen-go version: $PROTOC_GEN_GO_REV"
+PROTOC_GEN_GO_SHA256=$(nix-prefetch-git https://github.com/protocolbuffers/protobuf-go --rev "$PROTOC_GEN_GO_REV" | jq -r .hash)
```
Comment (dannykopping) [inline scripts/update-flake.sh:37]: @joobisb can we not use the `nix-command flakes` approach for this? My previous comment will apply to this, too (about compatibility).

Comment (dannykopping) [reply scripts/update-flake.sh:37]: Actually this is fine, I think.

**Review by dannykopping — APPROVED** [review]
Thanks for much for your contribution @joobisb!

**Comment by coadler** [comment]
@joobisb I'm seeing the following failure on my NixOS machine
```
$ nix develop
warning: Git tree '/home/colin/Projects/coder/coder' is dirty
error: builder for '/nix/store/wg14v4ylmjrf2krlxbk6a049nvc23zad-protoc-gen-go.drv' failed with exit code 1;
       last 10 log lines:
       > calling 'postUnpack' function hook '_updateSourceDateEpochFromSourceRoot'
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > calling 'preConfigure' function hook '_multioutConfig'
       > Running phase: buildPhase
       > evaling implicit 'preBuild' string hook
       > go: github.com/golang/protobuf@v1.5.0: Get "https://proxy.golang.org/github.com/golang/protobuf/@v/v1.5.0.info": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:33535->[::1]:53: read: connection refused
       > go: github.com/google/go-cmp@v0.5.5: Get "https://proxy.golang.org/github.com/google/go-cmp/@v/v0.5.5.info": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:54889->[::1]:53: read: connection refused
       > /nix/store/5r0df66ikad3xw06azlqvswcvncll8wa-stdenv-linux/setup: line 193: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/wg14v4ylmjrf2krlxbk6a049nvc23zad-protoc-gen-go.drv'.
error: 1 dependencies of derivation '/nix/store/z9ik49f82idmhi99ra9zvd3q38cqsg9s-nix-shell-env.drv' failed to build
```

On Linux, nix builds restrict network access. I was able to fix by deleting the following lines:

```diff
           subPackages = [ "cmd/protoc-gen-go" ];
           vendorHash = null;
-          proxyVendor = true;
-          preBuild = ''
-            export GOPROXY=https://proxy.golang.org,direct
-            go mod download
-          '';
-
```

Is there a reason this `preBuild` hook is necessary?

**Pull request merged**